pkg/river/encoding/block.go

```diff
@@ -0,0 +1,86 @@
+package encoding
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/grafana/agent/pkg/river/internal/rivertags"
+	"github.com/grafana/agent/pkg/river/internal/value"
+)
+
+// BlockField represents the json representation of a river block.
+type BlockField struct {
+	Field           `json:",omitempty"`
+	ID              string        `json:"id,omitempty"`
+	Label           string        `json:"label,omitempty"`
+	Body            []interface{} `json:"body,omitempty"`
+	attributeFields []*AttributeField
+	blockFields     []*BlockField
+}
+
+func newBlock(val value.Value, f rivertags.Field) (*BlockField, error) {
+	bf := &BlockField{}
+	return bf, bf.convertBlock(val, f)
+}
```

Comment:
Blocks are structural and are _not_ values (even though they can be encoded as them); this is the one place where we should be using the reflect API down the call chain.

(i.e., use reflect.Value here in `newBlock` and in `convertBlock`)